### ISSUE TITLE
puzzlefs linux kernel module

### ISSUE THREAD

**Issue by ariel-miculas** [anuvu/puzzlefs#78]
This issue will track the necessary changes required to get puzzlefs into the linux kernel.

**Comment by ariel-miculas** [comment]
The following dependencies are required for implementing a puzzlefs kernel module:
* the serde crate and its dependencies https://github.com/Rust-for-Linux/linux/pull/1007/commits
* the filesystem abstractions for rust in the linux kernel: https://github.com/Rust-for-Linux/linux/issues/1004 and https://rust-for-linux.zulipchat.com/#narrow/stream/384413-Filesystems/topic/Filesystem.20abstractions
* serde_cbor integration into kernel, potentially with the `alloc` feature if it can be supported: https://rust-for-linux.zulipchat.com/#narrow/stream/291565-Help/topic/Missing.20parts.20of.20alloc
* `this_error` integration into the kernel

Licensing discussion:
https://rust-for-linux.zulipchat.com/#narrow/stream/291565-Help/topic/Missing.20parts.20of.20alloc
https://docs.kernel.org/process/license-rules.html

**Comment by ariel-miculas** [comment]
List of missing rust features in the kernel, needed by puzzlefs:
* a [TryClone](https://docs.rs/fallible_collections/latest/fallible_collections/trait.TryClone.html) trait, use case: `try_clone()` a Vec
* a `TryFromIterator` together with a `try_collect` method, used extensively in puzzlefs and other crates such as [hex](https://docs.rs/hex/latest/hex/trait.ToHex.html).

**Comment by ariel-miculas** [comment]
First RFC patch series: https://lore.kernel.org/rust-for-linux/ZIO%2FOqxmFkTFzyDh@moria.home.lan/T/#t
Github branch: https://github.com/Rust-for-Linux/linux/compare/rust-next...ariel-miculas:puzzlefs_rfc_v1
Article in phoronix: https://www.phoronix.com/news/Experimental-Rust-PuzzleFS
Next steps towards upstreaming will be discussed with the `Rust for Linux` people on June 21st: https://rust-for-linux.zulipchat.com/#narrow/stream/384413-Filesystems/topic/Filesystem.20abstractions

**Comment by ariel-miculas** [comment]
v2 of the kernel driver: https://lore.kernel.org/all/20230726164535.230515-1-amiculas@cisco.com/
Github branch: https://github.com/ariel-miculas/linux/tree/puzzlefs_rfc_v2

**Comment by ariel-miculas** [comment]
There's a [new version](https://github.com/ariel-miculas/linux/tree/puzzlefs-ro-fs) of the puzzlefs driver based on Wedson's [read-only filesystem abstractions branch](https://github.com/wedsonaf/linux/commits/ro-fs). @tych0 FYI.

**Comment by tych0** [comment]
Nice, thanks for the update. Is your plan to archive this repo at some point?

I'm still finagling some time to work on PuzzleFS at work. I'm hopeful that after the first of the year I'll be able to lend some hours to the effort!

**Comment by tych0** [comment]
Fair enough, I didn't know if those were going to live in tools/ or something as well. Thanks!

**Comment by ariel-miculas** [comment]
First we need to convince people it is upstream-worthy.

**Comment by tych0** [comment]
Of course :)

**Comment by ariel-miculas** [comment]
We now have [a dedicated page](https://rust-for-linux.com/puzzlefs-filesystem-driver) under Rust for Linux.

**Comment by ariel-miculas** [comment]
The latest PuzzleFS patch series is [here](https://lore.kernel.org/all/20240516190345.957477-1-amiculas@cisco.com/). There have been discussions about PuzzleFS in [this email thread](https://lore.kernel.org/rust-for-linux/80cd0899-f14c-42f4-a0aa-3b8fa3717443@linux.alibaba.com/) and #114.
